File: lexer.py
from grammar import *
from decimal import Decimal

class Token(object):
	def __init__(self, token_type, value, line_num):
		self.type = token_type
		self.value = value
		self.cast()
		self.line_num = line_num

	def cast(self):
		if self.type == 'NUMBER':
			try:
				self.value = int(self.value)
			except ValueError:
				try:
					self.value = Decimal(self.value)
				except ValueError:
					pass

	def __str__(self):
		return 'Token(type={type}, value={value}, line_num={line_num})'.format(
			type=self.type,
			value=repr(self.value),
			line_num=self.line_num
		)

	__repr__ = __str__


class Lexer(object):
	def __init__(self, text):
		self.text = text
		self.pos = 0
		self.current_char = self.text[self.pos]
		self.char_type = None
		self.word = ''
		self.word_type = None
		self.line_num = 1

	def next(self):
		self.pos += 1
		if self.pos > len(self.text) - 1:
			self.current_char = None
			self.char_type = None
		else:
			self.current_char = self.text[self.pos]
			self.char_type = self.get_type(self.current_char)

	def reset_word(self):
		old_word = self.word
		self.word = ''
		self.word_type = None
		return old_word

	def peek(self, num):
		peek_pos = self.pos + num
		if peek_pos > len(self.text) - 1:
			return None
		else:
			return self.text[peek_pos]

	def preview_token(self, num=1):
		if num < 1:
			raise ValueError('num argument must be 1 or greater')
		next_token = None
		current_pos = self.pos
		current_char = self.current_char
		current_char_type = self.char_type
		current_word = self.word
		current_word_type = self.word_type
		for _ in range(num):
			next_token = self.get_next_token()
		self.pos = current_pos
		self.current_char = current_char
		self.char_type = current_char_type
		self.word = current_word
		self.word_type = current_word_type
		return next_token

	def skip_whitespace(self):
		if self.peek(-1) == '\n':
			raise SyntaxError('Only tab characters can indent')
		while self.current_char is not None and self.current_char.isspace():
			self.next()

	def skip_comment(self):
		while self.current_char != '\n':
			self.next()

	def eat_newline(self):
		self.reset_word()
		self.line_num += 1
		self.next()
		return Token('NEWLINE', '\\n', self.line_num - 1)

	def eat_indent(self):
		self.reset_word()
		self.next()
		return Token('INDENT', '\\t', self.line_num)

	@staticmethod
	def get_type(char):
		if char.isspace():
			return WHITESPACE
		if char == '#':
			return COMMENT
		if char == '\\':
			return ESCAPE
		if char in OPERATORS:
			return OPERATIC
		try:
			int(char)
			return NUMERIC
		except ValueError:
			return ALPHANUMERIC

	def get_next_token(self):
		if self.current_char is None:
			return Token('END', '', self.line_num)

		if self.current_char == '\n':
			return self.eat_newline()

		elif self.current_char == '\t':
			return self.eat_indent()

		if self.current_char.isspace():
			self.skip_whitespace()

		if self.current_char == '#':
			self.skip_comment()
			return self.eat_newline()

		if self.current_char == '"':
			self.next()
			while self.current_char != '"':
				if self.current_char == '\\' and self.peek(1) == '"':
					self.next()
				self.word += self.current_char
				self.next()
			self.next()
			return Token('STRING', self.reset_word(), self.line_num)

		if self.current_char == "'":
			self.next()
			while self.current_char != "'":
				if self.current_char == '\\' and self.peek(1) == "'":
					self.next()
				self.word += self.current_char
				self.next()
			self.next()
			return Token('STRING', self.reset_word(), self.line_num)

		if not self.char_type:
			self.char_type = self.get_type(self.current_char)
		if not self.word_type:
			self.word_type = self.char_type

		if self.word_type == ALPHANUMERIC:
			while self.char_type == ALPHANUMERIC or self.char_type == NUMERIC:
				self.word += self.current_char
				self.next()
			if self.word in KEYWORDS:
				return Token('KEYWORD', self.reset_word(), self.line_num)
			elif self.word in TYPES:
				return Token('TYPE', self.reset_word(), self.line_num)
			elif self.word in CONSTANTS:
				return Token('CONSTANT', self.reset_word(), self.line_num)
			else:
				return Token('NAME', self.reset_word(), self.line_num)

		if self.word_type == NUMERIC:
			while self.char_type == NUMERIC or self.current_char == '.':
				self.word += self.current_char
				self.next()
				if self.char_type == ALPHANUMERIC:
					raise SyntaxError('Variables cannot start with numbers')
			return Token('NUMBER', self.reset_word(), self.line_num)

		if self.word_type == OPERATIC:
			while self.char_type == OPERATIC:
				self.word += self.current_char
				self.next()
			return Token('OP', self.reset_word(), self.line_num)

		if self.char_type == ESCAPE:
			self.reset_word()
			self.next()
			line_num = self.line_num
			if self.current_char == '\n':
				self.line_num += 1
			self.next()
			return Token(ESCAPE, '\\', line_num)

		raise SyntaxError('Unknown character')

	def analyze(self):
		token = self.get_next_token()
		while token.type != 'END':
			yield token
			token = self.get_next_token()
		yield token


if __name__ == '__main__':
	lexer = Lexer(open('example.my').read())
	for t in lexer.analyze():
		if t.type == 'NEWLINE' or t.type == 'BEGIN':
			print(t)
		else:
			print(t, end=' ')

File: mythril/lexer/lexer.cpp
#include "lexer.h"
#include "../exceptions.h"
#include "my_grammar.h"
#include <algorithm>
#include <iostream>
#include <string>
using namespace std;

ostream &operator<<(ostream &out_stream, const Token &token) {
	map<TokenType, string> token_type_map = {
		{TokenType::TYPE, "TYPE"},
		{TokenType::NUMBER, "NUMBER"},
		{TokenType::STRING, "STRING"},
		{TokenType::OP, "OP"},
		{TokenType::CONSTANT, "CONSTANT"},
		{TokenType::NEWLINE, "NEWLINE"},
		{TokenType::INDENT, "INDENT"},
		{TokenType::KEYWORD, "KEYWORD"},
		{TokenType::ANON, "ANON"},
		{TokenType::NAME, "NAME"},
		{TokenType::END_OF_FILE, "EOF"},
	};

	map<char, string> whitespace_map = {
		{'\t', "\\t"},
		{'\n', "\\n"},
		{'\v', "\\v"},
		{'\f', "\\f"},
		{'\r', "\\r"},
	};

	string value = token.value;

	if (isspace((int) value[0])) {
		value = '\'' + whitespace_map[(int) value[0]] + '\'';
	}
	else if (token.type == TokenType::END_OF_FILE) {
		value = "'EOF'";
	}
	else if (token.type != TokenType::NUMBER) {
		bool in_str = value.find('\'') != string::npos;
		if (in_str) {
			value = '"' + value + '"';
		}
		else {
			value = '\'' + value + '\'';
		}
	}
	out_stream
		<< "Token(type="
		<< token_type_map[token.type]
		<< ", value="
		<< value
		<< ", line_num="
		<< token.line_num
		<< ", indent_level="
		<< token.indent_level
		<< ")";
	return out_stream;
}

Lexer::Lexer(string &text) {
	_text = text;
	current_char = _text[pos];
	char_type = CharType::NONE;
	word_type = CharType::NONE;
}

Token Lexer::eof() {
	return Token{
		.indent_level=_indent_level,
		.line_num=_line_num,
		.type=TokenType::END_OF_FILE,
		.value=word,
	};
}

void Lexer::next_char() {
	pos += 1;
	if (pos > _text.length() - 1) {
		current_char = '\0';
		char_type = CharType::NONE;
	}
	else {
		current_char = _text[pos];
		char_type = get_type(current_char);
	}
}

string Lexer::reset_word() {
	string old_word = word;
	word = "";
	word_type = CharType::NONE;
	return old_word;
}

char Lexer::peek(int num) {
	int peek_pos = pos + num;
	if (peek_pos > _text.length() - 1 && peek_pos < 0) {
		return '\0';
	}
	else {
		return _text[peek_pos];
	}
}

void Lexer::skip_whitespace() {
	if (peek(-1) == '\n') {
		throw SyntaxError(
			"Only tab characters can indent",
			_line_num,
			current_char
		);
	}
	while (current_char != EOF && current_char == ' ') {
		next_char();
		reset_word();
	}
}

CharType Lexer::get_type(char a_char) {
	string a_str(1, a_char);
	if (a_char == ' ') {
		return CharType::WHITEPSACE;
	}
	if (a_char == '\n') {
		return CharType::NEW_LINE;
	}
	if (a_char == '#') {
		return CharType::COMMENT;
	}
	if (a_char == '\\') {
		return CharType::ESCAPE;
	}
	bool in_ops = find(begin(OPERATORS), end(OPERATORS), a_str) != end(OPERATORS);
	if (in_ops) {
		return CharType::OPERATIC;
	}
	if (isdigit(a_char)) {
		return CharType::NUMERIC;
	}
	return CharType::ALPHANUMERIC;
}

Token Lexer::eat_newline() {
	reset_word();
	Token token = {
		.type=TokenType::NEWLINE,
		.value="\n",
		.line_num=_line_num,
		.indent_level=_indent_level
	};
	_indent_level = 0;
	_line_num += 1;
	next_char();
	return token;
}

Token Lexer::eat_string() {
	char quote_char = current_char;
	next_char();
	while (current_char != quote_char) {
		if (current_char == '\\' && peek(1) == quote_char) {
			next_char();
		}
		word += current_char;
		next_char();
	}
	next_char();
	return make_token(TokenType::STRING);
}

void Lexer::skip_indent() {
	while (current_char != EOF && current_char == '\t') {
		reset_word();
		_indent_level += 1;
		next_char();
	}
}

void Lexer::skip_comment() {
	while (current_char != '\n') {
		next_char();
		if (current_char == EOF) {
			return;
		}
	}
	eat_newline();
	if (current_char == '#') {
		skip_comment();
	}
}

Token Lexer::preview_token(int num) {
	if (num < 1) {
		throw invalid_argument("num argument must be 1 or greater");
	}

	Token next_token = {};
	int _pos = pos;
	char _current_char = current_char;
	CharType _char_type = char_type;
	string _word = word;
	CharType _word_type = word_type;
	int __line_num = _line_num;
	int __indent_level = _indent_level;

	while (num >= 0) {
		next_token = get_next_token();
		num -= 1;
	}

	pos = _pos;
	current_char = _current_char;
	char_type = _char_type;
	word = _word;
	word_type = _word_type;
	_line_num = __line_num;
	_indent_level = __indent_level;

	return next_token;
}

Token Lexer::make_token(TokenType type) {
	return Token{
		.type = type,
		.value = reset_word(),
		.line_num = _line_num,
		.indent_level = _indent_level,
	};
}

Token Lexer::get_next_token() {
	if (current_char == EOF || current_char == '\0') {
		return eof();
	}

	if (current_char == '\n') {
		return eat_newline();
	}

	if (current_char == '\t') {
		skip_indent();
	}

	if (current_char == ' ') {
		skip_whitespace();
	}

	if (current_char == '#') {
		skip_comment();
		return get_next_token();
	}

	string current_char_str(1, current_char);
	bool is_quote_char = find(
		begin(QUOTES),
		end(QUOTES),
		current_char_str
	) != end(QUOTES);

	if (is_quote_char) {
		return eat_string();
	}

	if (char_type == CharType::NONE) {
		char_type = get_type(current_char);
	}
	if (word_type == CharType::NONE) {
		word_type = char_type;
	}

	if (word_type == CharType::OPERATIC) {
		while (char_type == CharType::OPERATIC) {
			word += current_char;
			next_char();

			string current_char_str(1, current_char);
			bool cc_in_single_op = find(
				begin(SINGLE_OPERATORS),
				end(SINGLE_OPERATORS),
				current_char_str
			) != end(SINGLE_OPERATORS);
			bool word_in_single_op = find(
				begin(SINGLE_OPERATORS),
				end(SINGLE_OPERATORS),
				word
			) != end(SINGLE_OPERATORS);

			if (cc_in_single_op || word_in_single_op) {
				word_type = CharType::NONE;
				break;
			}
		}
		return make_token(TokenType::OP);
	}

	if (word_type == CharType::ALPHANUMERIC) {
		while (char_type == CharType::ALPHANUMERIC || char_type == CharType::NUMERIC) {
			word += current_char;
			next_char();
		}

		bool in_op = find(begin(OPERATORS), end(OPERATORS), word) != end(OPERATORS);

		if (in_op) {
			string pt = preview_token(1).value;
			bool word_in_mw_op = find(
				begin(MULTI_WORD_OPERATORS),
				end(MULTI_WORD_OPERATORS),
				word
			) != end(MULTI_WORD_OPERATORS);
			bool pt_in_mw_op = find(
				begin(MULTI_WORD_OPERATORS),
				end(MULTI_WORD_OPERATORS),
				pt
			) != end(MULTI_WORD_OPERATORS);

			if (word_in_mw_op && pt_in_mw_op) {
				next_char();
				word += ' ';
				while (
					char_type == CharType::ALPHANUMERIC
					|| char_type == CharType::NUMERIC
				) {
					word += current_char;
					next_char();
				}
			}
			return make_token(TokenType::OP);
		}

		bool in_keywords = find(begin(KEYWORDS), end(KEYWORDS), word) != end(KEYWORDS);
		bool in_types = find(begin(TYPES), end(TYPES), word) != end(TYPES);
		bool in_constants = find(
			begin(CONSTANTS),
			end(CONSTANTS),
			word
		) != end(CONSTANTS);

		if (in_keywords) {
			string pt = preview_token(1).value;
			bool word_in_mw_keywords = find(
				begin(MULTI_WORD_KEYWORDS),
				end(MULTI_WORD_KEYWORDS),
				word
			) != end(MULTI_WORD_KEYWORDS);
			bool pt_in_mw_keywords = find(
				begin(MULTI_WORD_KEYWORDS),
				end(MULTI_WORD_KEYWORDS),
				pt
			) != end(MULTI_WORD_KEYWORDS);

			if (word_in_mw_keywords && pt_in_mw_keywords) {
				next_char();
				word += ' ';
				while (
					char_type == CharType::ALPHANUMERIC
					|| char_type == CharType::NUMERIC
				) {
					word += current_char;
					next_char();
				}
			}
			return make_token(TokenType::KEYWORD);
		}
		else if (in_types) {
			return make_token(TokenType::TYPE);
		}
		else if (in_constants) {
			return make_token(TokenType::CONSTANT);
		}
		else {
			return make_token(TokenType::NAME);
		}
	}

	if (word_type == CharType::NUMERIC) {
		while (
			char_type == CharType::NUMERIC || (current_char == '.' && peek(1) != '.')
		) {
			word += current_char;
			next_char();
			if (char_type == CharType::ALPHANUMERIC) {
				throw SyntaxError(
					"Variables cannot start with numbers",
					_line_num,
					current_char
				);
			}
		}
		string value = reset_word();
		MythrilType value_type;
		bool in_str = value.find('.') != string::npos;
		if (in_str) {
			value_type = MythrilType::DEC;
		}
		else {
			value_type = MythrilType::INT;
		}
		Token token = make_token(TokenType::NUMBER);
		token.value_type = value_type;
		return token;
	}

	throw SyntaxError("Unknown Character", _line_num, current_char);
}

vector<Token> Lexer::analyze() {
	Token token = get_next_token();
	tokens.push_back(token);
	while (token.type != TokenType::END_OF_FILE) {
		token = get_next_token();
		tokens.push_back(token);
	}
	return tokens;
}
